### DISCUSSION TITLE
Type hint/generics support

### DISCUSSION THREAD

**Discussion by molotch** [jasondelaat/pymonad#19] [Q&A]
I couldn't find any support for type hints/generics in the documentation. Is it possible to use typed Maybes and get back typed functions when using currying for instance?

I tried curry a function with type hints but Pylance have no clue about the returned function.

**Comment by jasondelaat** [comment]
So the short honest answer is: I'm not sure but I'll look into it.
I'm not familiar with Pylance so I'm not sure about that specifically. I do know there are systems for static type checking, like PyType, and I think there are libraries that allow for checking/enforcing type hints at run time so it must be possible to extract the typing information which leads me to believe that it _might_ be possible to add typing information to curried functions dynamically. 

For static checkers like PyType, I think you'd need to modify those tools specifically to know about curried functions and how to infer their types and I have no idea what would be involved in that. 

I think I've hinted everything in pymonad correctly but honestly, I don't really know much about the internals for the typing/hinting system so I'm afraid I don't have a good answer for you at the moment. But, as I said, I'll look into it and try to keep you posted. Fair warning though, work is pretty busy for me at the moment so it'll probably be at least a couple weeks before I have the time to really dig into it.

**Reply by molotch** [reply]
You're probably right, it's down to the tools. Pylance is using Microsofts open sourced Pyright for type checking. Think I'll post a suggestion on their board. Using partial from functools suffers from the same problem.

**Reply by jasondelaat** [reply]
I hadn't considered partial but yeah, makes sense that it has similar problems.
